exp/operator/controllers/coreprovider_controller.go

```diff
@@ -19,31 +19,101 @@ package controllers
 }
 
+func (r *CoreProviderReconciler) SetupWithManager(mgr ctrl.Manager, options controller.Options) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&operatorv1.CoreProvider{}).
+		WithOptions(options).
+		Complete(r)
+}
+
 // +kubebuilder:rbac:groups=operator.cluster.x-k8s.io,resources=coreproviders,verbs=get;list;watch;create;update;patch;delete
 // +kubebuilder:rbac:groups=operator.cluster.x-k8s.io,resources=coreproviders/status,verbs=get;update;patch
 
-func (r *CoreProviderReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+func (r *CoreProviderReconciler) Reconcile(ctx context.Context, req ctrl.Request) (_ ctrl.Result, reterr error) {
 	log := ctrl.LoggerFrom(ctx)
 	log.V(4).Info("Reconcile CoreProvider")
 
+	coreProvider := &operatorv1.CoreProvider{}
+
+	if err := r.Client.Get(ctx, req.NamespacedName, coreProvider); err != nil {
+		if apierrors.IsNotFound(err) {
+			// Object not found, return. Created objects are automatically garbage collected.
+			// For additional cleanup logic use finalizers.
+			return ctrl.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return ctrl.Result{}, err
+	}
+
+	// Initialize the patch helper
+	patchHelper, err := patch.NewHelper(coreProvider, r.Client)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	defer func() {
+		// Always attempt to patch the object and status after each reconciliation.
+		if err := patchHelper.Patch(ctx, coreProvider); err != nil {
+			reterr = kerrors.NewAggregate([]error{reterr, err})
+		}
+	}()
+
+	// Ignore deleted CoreProvider, this can happen when foregroundDeletion
+	// is enabled
+	if !coreProvider.DeletionTimestamp.IsZero() {
+		return ctrl.Result{}, nil
+	}
+
+	return r.reconcile(ctx, coreProvider)
+}
+
+func (r *CoreProviderReconciler) reconcile(ctx context.Context, coreProvider *operatorv1.CoreProvider) (ctrl.Result, error) {
+	// Run preflight checks to ensure that core provider can be installed properly
+	if result, err := r.preflightChecks(ctx, coreProvider); err != nil || !result.IsZero() {
+		return result, err
+	}
+
 	return ctrl.Result{}, nil
 }
 
-func (r *CoreProviderReconciler) SetupWithManager(mgr ctrl.Manager, options controller.Options) error {
-	return ctrl.NewControllerManagedBy(mgr).
-		For(&operatorv1alpha4.CoreProvider{}).
-		WithOptions(options).
-		Complete(r)
+func (r *CoreProviderReconciler) preflightChecks(ctx context.Context, coreProvider *operatorv1.CoreProvider) (ctrl.Result, error) {
+	log := ctrl.LoggerFrom(ctx)
+
+	coreProviderList := &operatorv1.CoreProviderList{}
+
+	if err := r.Client.List(ctx, coreProviderList, &client.ListOptions{}); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	if len(coreProviderList.Items) > 1 {
+		log.V(4).Info("More than one core provider exists in the cluster. Only one is allowed")
+		conditions.MarkFalse(coreProvider,
+			operatorv1.PreflightCheckCondition,
+			operatorv1.MoreThanOneExistsReason,
+			v1alpha4.ConditionSeverityInfo,
```

Comment:
Why is the condition severity info? I would have expected error or warning.

Reply:
👍